Add S3Cache

alexjurkiewicz opened the pull request:
Is there interest in merging a cache backed with S3? If so, I will complete work on this PR.

At the moment it's incomplete:

- [x] Custom bucket
- [x] Key prefix
- [x] `has`, `get`, `set`
- [x] `add`
- [x] `delete`
- [x] `clear`
- [x] timeout
- [x] `delete_many`
- [ ] `inc`, `dec`
- [x] `get_dict`, `get_many`
- [x] `set_many`
- [x] Test

ThiefMaster commented:
Not sure if that makes much sense - a cache is usually meant to be something very fast, and I don't think S3 is a great choice for that. AFAIK it's more something people use to put bigger files etc, but not the kind of data snippets one usually puts in a cache..

alexjurkiewicz commented:
The performance of S3 is probably better than you expect. Keys can handle
thousands of reads and writes per second. Time to first byte is 100-200ms.
Both these data points are from here:
https://docs.aws.amazon.com/AmazonS3/latest/dev/optimizing-performance.html

S3 is in many ways similar to the filesystem cache. It would be useful in
situations where objects are relatively large and the cache miss time is
relatively high. For example, systems which deal with converting media
files, such as on-demand image resizing or audio re-encoding from master
copies.

S3 works in places where the filesystem cache does not, for example in AWS
compute resources. Most particularly Lambda, but also EC2/ECS/EKS, where
the filesystem cache could only work via EFS.

On Tue, 4 Jun 2019 at 20:32, Adrian <notifications@github.com> wrote:

> Not sure if that makes much sense - a cache is usually meant to be
> something very fast, and I don't think S3 is a great choice for that. AFAIK
> it's more something people use to put bigger files etc, but not the kind of
> data snippets one usually puts in a cache..
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/pallets/cachelib/pull/6?email_source=notifications&email_token=AAC4U5PWQRGWI3HJ5VFBGHTPYZACJA5CNFSM4HS2MT22YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODW4EQTQ#issuecomment-498616398>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAC4U5JXW6IVNNEIEI5OVPLPYZACJANCNFSM4HS2MT2Q>
> .
>

ThiefMaster commented:
ok fair enough, for large cached objects it makes more sense

alexjurkiewicz commented:
This is ready for review. Caveats / differences from base cache implementation:

1. I didn't implement `inc`/`dec` because it seemed pointless
2. TTL is implemented by setting the Expires header. I added a parameter to automatically clean up stale data, but real production use should probably rely on a S3 bucket lifecycle rule to clean up old objects.

I've added pytest for S3Cache but it's not hooked up to tox or Travis or what have you. I don't think that's in scope for this PR. You can run the tests with `python -m pytest`.

alexjurkiewicz commented:
Also, this PR only works on Python 3. I assume that's fine? Also, it would be nice to bump the version of this package to 1.0.0 before merging this PR, and merge it as 1.1.0 (so the version numbers are useful).

alexjurkiewicz commented:
Finally, testing cache expiry could be sped up with https://github.com/spulec/freezegun (I won't do it as part of this PR, this is just a note for the future).

bfaludi commented:
@ThiefMaster @alexjurkiewicz Wondering would it be possible to merge this? :) 

Superset's documentation is mentioning an `s3cache` library but that package is not compatible with the werkzeug's abstract class anymore. Since most of the caching methods were moved here, It would be awesome to have S3 cache option back without hacking around this functionality. :)

https://superset.incubator.apache.org/installation.html?highlight=cache

RonnyPfannschmidt commented on cachelib/s3.py:
```diff
@@ -0,0 +1,191 @@
+except ImportError:  # pragma: no cover
+    import pickle
+
+try:
```
this bit has to go in the `__init__`, else the complete package fails without boto3 installed

alexjurkiewicz commented:
Sorry, someone else will have to pick up the torch here, I'm not interested
in working further on it.

On Thu, 12 Mar 2020, 5:15 pm Ronny Pfannschmidt, <notifications@github.com>
wrote:

> *@RonnyPfannschmidt* requested changes on this pull request.
> ------------------------------
>
> In cachelib/s3.py
> <https://github.com/pallets/cachelib/pull/6#discussion_r391419340>:
>
> > @@ -0,0 +1,191 @@
> +# -*- coding: utf-8 -*-
> +import datetime
> +
> +try:
> +    import cPickle as pickle
> +except ImportError:  # pragma: no cover
> +    import pickle
> +
> +try:
>
> this bit has to go in the __init__, else the complete package fails
> without boto3 installed
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pallets/cachelib/pull/6#pullrequestreview-373300032>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAC4U5IMBT7FIP6YY3KHFC3RHB4YLANCNFSM4HS2MT2Q>
> .
>

lepture commented:
I don't think S3 is a good choice for cache.
